### PR TITLE
Migrate to full css-variables theming with IThemes provider

### LINKED ISSUE #9246: server/apps/theming – standard nc css gets loaded before theming app css
### Steps to reproduce
1. load nc without cached browser files   
2. theming app loads a ~1,8Mb background picture  
3. internet connection with 1,5Mb/s download

### Expected behaviour
- theming app's css should be loaded FIRST.
- standard nc css (nc logo, blue background colour) should be loaded 	IN NO CASE!

### Actual behaviour
1. standard nc css (nc logo, blue background colour) gets loaded (more or less completely)
2. after 0.3-1.5 sec it gets replaced with the desired theming app's css (black background, background picture, theming app logo)

### Server configuration

**Operating system**:
debian 9

**Web server:**
apache2

**Database:**
mysql (MariaDB)

**PHP version:**
7.0.27

**Nextcloud version:**
13.0.1

### LINKED ISSUE #14059: Changing values of the theming app restyles without accessibility settings
1. Enable dark mode
2. Enable dyslexia font
3. Go to /index.php/settings/admin/theming
4. Change any value
5. Be blinded by the white page :dark_sunglasses:  

cc @juliushaertl @skjnldsv

### PR DESCRIPTION
**Pull request by nickvergessen**
This is a first start towards a proper theme management system.
That should help us do the following:
- Get a cleaner and more flexible approach towards Theming in general
- Allow external apps to provide themes in the future
- Allow the use of media-queries to select a specific theme automatically
- Deprecate the old scssCacher method
- Migrate away from scss #31743 
- Allow for much better performances in design generation and per-user design caching
- Fix most of the guest/public/login pages that were not themed
- Move to a declarative theming approach rather than imperative (apps handle their themes compatibility instead of having the theme provide custom css rules)

## Steps
- [x] Implement base `ITheme` provider (private for now)
- [x] Implement theme management backend
- [x] Migrate Theme selection UI from accessibility to theming
- [x] Adapt UI
- [x] Implement instant design switch
- [x] Remove accessibility

## How to review
- Test the various ways themes can be changed per-user
  - From the user theming settings
  - With media queries
- Try to find visual glitches
- Experiment with Admin theming changes and colours/edits and see if everything behaves properly
  - Settings changes and preview
  - Public pages
  - Standard logged-in pages
  - Login page

## Follow-ups
- [ ] Remove scss support https://github.com/nextcloud/server/pull/31743
- [ ] Fix icons generation
  - [ ] Compile with npm
  - [ ] Implement theme-switch icon colour revert (?)
- [ ] Fix potential leftovers and deploy magic variables like
  - `--background-invert-if-bright` => invert the element if the main background colour is bright
  - `--primary-invert-if-bright` => invert the element if the primary colour is too bright
- [ ] Make ITheme api public (?)
____
Peek 21-04-2022 09-37

Closes #31746

### CODE REVIEW

File: apps/theming/lib/Controller/ThemingController.php [hunk 1]
```diff
@@ -341,25 +309,29 @@ public function getImage(string $key, bool $useSvg = true) {
...
-		$appPath = $this->appManager->getAppPath('theming');
-
-		/* SCSSCacher is required here
-		 * We cannot rely on automatic caching done by \OC_Util::addStyle,
-		 * since we need to add the cacheBuster value to the url
-		 */
-		$cssCached = $this->scssCacher->process($appPath, 'css/theming.scss', 'theming');
-		if (!$cssCached) {
+	public function getThemeVariables(string $themeId, bool $plain) {
+		$theme = $this->themesService->getThemes()[$themeId];
```
Comment (come-nc) [inline apps/theming/lib/Controller/ThemingController.php:314]:
```suggestion
		$theme = $this->themesService->getThemes()[$themeId] ?? null;
```

File: apps/theming/lib/Service/ThemeInjectionService.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		$this->themesService = $themesService;
+		$this->defaultTheme = $defaultTheme;
+	}
+
+	public function injectHeaders() {
+		$themes = $this->themesService->getThemes();
+		$defaultTheme = $themes[$this->defaultTheme->id];
+		$mediaThemes = array_filter($themes, function($theme) {
+			// Check if the theme provides a media query
+			return !!$theme->getMediaQuery();
```
Comment (come-nc) [inline apps/theming/lib/Service/ThemeInjectionService.php:43]:
```suggestion
			return (bool)$theme->getMediaQuery();
```

Comment (come-nc) [reply apps/theming/lib/Service/ThemeInjectionService.php:43]: Or even
```suggestion
			return ($theme->getMediaQuery() !== '');
```

File: apps/theming/lib/Service/ThemesService.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	public function __construct(DefaultTheme $defaultTheme,
+								DarkTheme $darkTheme,
+								DarkHighContrastTheme $darkHighContrastTheme,
+								HighContrastTheme $highContrastTheme) {
+		// Register themes
+		$this->themesProviders = [
+			$defaultTheme->id			=> $defaultTheme,
+			$darkTheme->id				=> $darkTheme,
+			$darkHighContrastTheme->id	=> $darkHighContrastTheme,
+			$highContrastTheme->id		=> $highContrastTheme,
```
Comment (come-nc) [inline apps/theming/lib/Service/ThemesService.php:45]:
```suggestion
			$defaultTheme->getId()			=> $defaultTheme,
			$darkTheme->getId()				=> $darkTheme,
			$darkHighContrastTheme->getId()	=> $darkHighContrastTheme,
			$highContrastTheme->getId()		=> $highContrastTheme,
```

**Review by come-nc — DISMISSED** [review]

File: apps/theming/lib/Controller/ThemingController.php [hunk 4]
```diff
@@ -341,25 +309,29 @@ public function getImage(string $key, bool $useSvg = true) {
...
+		if (!$theme) {
 			return new NotFoundResponse();
 		}
 
+		// Generate variables
+		$variables = '';
+		foreach ($theme->getCSSVariables() as $variable => $value) {
+			$variable .= "$variable:$value; ";
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/theming/lib/Controller/ThemingController.php]: Not in css no :)
That would be invalid and discarded

**Comment by szaimen** [comment]
@skjnldsv could you please clarify which apps/parts of Nextcloud should be checked for problems when reviewing this PR?
Thank you :)

**Review by CarlSchwan — APPROVED** [review]
The code looks fine to me

I would propose to merge it sooner than later even if there are some small issues that can be fixed in follow-up as we are in the beginning of a release cycle.

**Comment by skjnldsv** [comment]
> The code looks fine to me
> 
> I would propose to merge it sooner than later even if there are some small issues that can be fixed in follow-up as we are in the beginning of a release cycle.

My thoughts as well.
I added some phpunit tests, fixed the old ones and psalm.
There will most likely be quite a few follow-ups, but that's a good first step.

**Comment by skjnldsv** [comment]
@juliushaertl there will most likely also be stuff I missed on the old magic Theming was doing. I removed most that I could, but I am sure I missed a few. Any help welcome on understanding anything that might be missing now. :pray:

**Comment by skjnldsv** [comment]
> @skjnldsv could you please clarify which apps/parts of Nextcloud should be checked for problems when testing this PR? Thank you :)

I'll edit the main post for review requirements :)

**Comment by szaimen** [comment]
> > @skjnldsv could you please clarify which apps/parts of Nextcloud should be checked for problems when testing this PR? Thank you :)
> 
> I'll edit the main post for review requirements :)

thanks :)

**Comment by szaimen** [comment]
There are many things that are broken and not working correctly. most important are probably the missing logo on the login screen the the default light mode clashing with the prefer color scheme media query.

<details>
<summary>click to expand</summary>


a few things are broken:
- [ ] missing login logo: image
- [ ] the text is more grey than before: image
- [ ] this text too: image
- [ ] should hovering be dark in light mode? image
- [ ] same: image
- [ ] also: image
- [ ] missing icons, also black popup in light mode? image
- [ ] in general it seems to respect my dark mode system setting but not for all elements
- [ ] it is strange that the default them will use the media query image
- [ ] the dyslexia font does not work for me: image
- [ ] viewer buuttons in light mode with high contrast does not respect this: image
- [ ] icons in dark with high contrast dont look correct: image
- [ ] also in dark mode the icons are not correct: image
- [ ] in dark mode font does not look correct: image
- [ ] opening public links breaks the text color again: image
- [ ] same for the text app: image
- [ ] I could go on...


</details>

**Review by szaimen — DISMISSED** [review]
There are many things that are broken and not working correctly. most important are probably the missing logo on the login screen, the default light mode clashing with the prefer color scheme media query and dyslexia not working. Also the theming logo and picture options dont seem to work as expected.

<details>
<summary>click to expand</summary>


a few things are broken:
- [ ] missing login logo: image
- [ ] the text is more grey than before: image
- [ ] this text too: image
- [ ] should hovering be dark in light mode? image
- [ ] same: image
- [ ] also: image
- [ ] missing icons, also black popup in light mode? image
- [ ] in general it seems to respect my dark mode system setting but not for all elements
- [ ] it is strange that the default them will use the media query image
- [ ] the dyslexia font does not work for me: image
- [ ] viewer buuttons in light mode with high contrast does not respect this: image
- [ ] icons in dark with high contrast dont look correct: image
- [ ] also in dark mode the icons are not correct: image
- [ ] in dark mode font does not look correct: image
- [ ] opening public links breaks the text color again: image
- [ ] same for the text app: image
- [ ] These options don't seem to work as expected: image

- [ ] I could go on...


</details>


<details>
<summary>For my own testing</summary>

```
docker run -it \
-e SERVER_BRANCH=theming-providers \
-e VIEWER_BRANCH=master \
-e TEXT_BRANCH=master \
-p 8443:443 \
-e TRUSTED_DOMAIN=192.168.146.128 \
--name nextcloud-easy-test \
ghcr.io/szaimen/nextcloud-easy-test:latest
```

</details>

**Comment by skjnldsv** [comment]
> a few things are broken

Fixed the login

The rest will be fixed in followups :)

**Comment by szaimen** [comment]
I'll retest in a second 👍

**Review by szaimen — COMMENTED** [review]
the dashboard app is still a bit broken in dark mode: 
image

and the files app too:
image
image

Also dyslexia still does not work for me, don't know why...

and this is still not renamed: 
image


But okay to me to do this in a follow-up

**Comment by szaimen** [comment]
We should add this to breaking change of 25 (e.g. that the accessibility app is now included in theming...)

**Comment by skjnldsv** [comment]
> But okay to me to do this in a follow-up

Will do! :rocket:

**Review by skjnldsv — APPROVED** [review]
Will address design improvements in followups

**Comment by skjnldsv** [comment]
integration-setup-features failure unrelated

**Comment by szaimen** [comment]
@skjnldsv Did you find a fix for the dyslexic prpblem?

**Comment by skjnldsv** [comment]
> @skjnldsv Did you find a fix for the dyslexic problem?

Yes, will add as followup, just using the generateUrl should be enough

**Comment by skjnldsv** [comment]
> @skjnldsv something to look into it next week, but I think this broke one integration test

I wondered this too, but I cannot reproduce locally :thinking:

**Comment by zak39** [comment]
Is it possible to declare sccs in a Vue template ?

**Comment by skjnldsv** [comment]
> Is it possible to declare sccs in a Vue template ?

Yes, we use scss in vue apps almost everywhere :)

**Pull request merged**